### PR TITLE
Site Editor: Navigation templates

### PR DESCRIPTION
**Pull request by david-szabo97**
## Description
A couple of things done in this PR to match the draft https://github.com/WordPress/gutenberg/issues/23939#issuecomment-698855666)

- Root menu is now `Themes`
- Deleted `Current Theme` group
- Deleted `Current Template` group
- Templates are now in a separate menu (`TemplatesMenu`)
- Template Parts are now in a separate menu (`TemplatePartsMenu`)
- Page templates are now in a separate menu (`TemplatesPagesMenu`)
- Post templates are now in a separate menu (`TemplatesPostsMenu`)
- Moved preview panel to the root and used slot-fill (had to do this because of `overflow:hidden` on the `Navigation` component)

## How has this been tested?
* `yarn wp-env start`
* Enable FSE
* Open Site Editor
* Click site logo to toggle the navigation sidebar

## Screenshots 
image
image
image

## Types of changes
New feature

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/edit-site/src/components/left-sidebar/navigation-panel/current-theme.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+} from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import ThemePreview from './theme-preview';
+
+export default function CurrentTheme() {
```
Comment (david-szabo97) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/current-theme.js:17]: Extracted Theme group to its own component

**Comment by david-szabo97** [comment]
Looking forward ideas about how we should handle empty menus. For example `Pages` item should navigate to `TemplatesPagesMenu` but it's empty so we skip rendering. I wonder what would be the best way to hide the `Pages` item when the menu is empty. Maybe we should implement this in the component?

**Comment by vindl** [comment]
> Extracted Current Theme group as a separate component (CurrentTheme ; we need to come up with a better name)

I don't think we need to show this in the sidebar anymore. At least I haven't seen it in any of latest design iterations. It might make more sense in Document Settings dropdown maybe.

Also I don't think we need a Current Template section. That should be obvious from active item and title in the header?

**Comment by vindl** [comment]
> Looking forward ideas about how we should handle empty menus. For example Pages item should navigate to TemplatesPagesMenu but it's empty so we skip rendering. I wonder what would be the best way to hide the Pages item when the menu is empty. Maybe we should implement something in the Navigation component?

It's probably fine to leave it as a placeholder for now until we add more items for it. And yes, this should be a separate issue for navigation component and not handled in this PR.

**Comment by david-szabo97** [comment]
> Also I don't think we need a Current Template section. That should be obvious from active item and title in the header?

I left it there so we can overwrite templates. I think it will be replaced by the + button later 🤔

**Comment by vindl** [comment]
> I left it there so we can overwrite templates. I think it will be replaced by the + button later 🤔

In that I case I think it's better to introduce it when we add the +/overwrite functions then?

**Comment by vindl** [comment]
cc @ockham since you worked on template resolution before :)

File: lib/edit-site-page-templates.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * REST endpoint for exposing theme's templates.
+ *
+ * @package gutenberg
+ */
+
+function gutenberg_edit_site_get_page_templates() {
```
Comment (noahtallen) [inline lib/edit-site-page-templates.php:8]: I'm not totally sure we need this. Does this existing code help?

https://github.com/WordPress/gutenberg/blob/848b4934242512ab3909e9851842dfac4732512a/lib/templates.php#L179-L201

There is another example here which is used for edit site export:

https://github.com/WordPress/gutenberg/blob/848b4934242512ab3909e9851842dfac4732512a/lib/edit-site-export.php#L21-L38

In each of those cases, I think we go through each of the available template types, and try to resolve templates for them, rather than looking through the theme files. IIRC, that code also handles creating autodrafts of templates which don't exist.

I might be wrong though, it's a confusing part of the codebase :)

Comment (david-szabo97) [reply lib/edit-site-page-templates.php:8]: My code is putting together an array of file paths of "physical template files". So these aren't stored as a `wp_template` in the database so we can't request it with REST API.

`filter_rest_wp_template_query` to utilise this, we need an actual template type that we can query. But we don't know those. That's why we need this little endpoint I created.

`auto-drafts` are created once the template is selected for editing.

Comment (Copons) [reply lib/edit-site-page-templates.php:8]: I'm trying to wrap my head around all of this, so apologize in advance if I'm missing the point, or saying something really dumb.

As far as I understand, these autodrafts of templates and template parts are created within the [`rest_{$this->post_type}_query`](https://developer.wordpress.org/reference/hooks/rest_this-post_type_query/) filters, which runs on `wp_template` and `wp_template_part` REST requests.

The reasoning behind running it every time is to check if a file has changed. E.g.:
https://github.com/WordPress/gutenberg/blob/759729cb22774acba6284045a3f2555c15fe16d4/lib/template-loader.php#L307-L310

I guess it makes sense, and it prevents us from, for example, creating the autodrafts once on theme activation.

Now, when we call this new `/page-templates` endpoint (btw: do we need one for template parts as well?), what happens with elements with a _file_, but without an _entity_ (aka without an autodraft)?
Are they going to be created as a side effect of `getEntityRecords`?
In other words, does `getEntityRecords` call a request on the post type, triggering the filter which creates the autodrafts needed?

Otherwise, we'd basically get a file name.
What do we do with it? 🤔

File: lib/template-loader.php [hunk 3]
```diff
@@ -89,6 +89,16 @@ function get_template_hierarchy( $template_type ) {
...
+	if ( isset( $_GET['_wp-find-template'] ) ) {
+		if ( ! empty( $_GET['_wp-find-template-type'] ) ) {
+			$type = $_GET['_wp-find-template-type'];
+		}
+
+		if ( ! empty( $_GET['_wp-find-template-templates'] ) ) {
+			$templates = $_GET['_wp-find-template-templates'];
+		}
+	}
+
```
Comment (Copons) [inline lib/template-loader.php:101]: What would be the purpose of this?

**Comment by Copons** [comment]
I've noticed a wonky behaviour when changing from template to template part and vice versa.
E.g.
- Open the Site Editor: it's showing the `index` template.
- In the sidebar, navigate to template parts and select `header`.

The editor's `EntityProvider` changes from `wp_template` to `wp_template_part` (at least I think that's the cause), causing a full remount (again: I think?) of the sidebar, which loses the `activeMenu` value and reverts to `root`.
Now, the active item is still correct, but the sudden change of active menu is definitely odd.

I think we should sort this out, although we might need to do some changes to the Navigation component.
I haven't fully investigated this yet, so I might very well be missing something obvious.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js [hunk 4]
```diff
@@ -1,80 +1,101 @@
...
+			template: templateId
+				? getEntityRecord( 'postType', 'wp_template', templateId )
+				: null,
+			templatePart: templatePartId
+				? getEntityRecord(
+						'postType',
+						'wp_template_part',
+						templatePartId
+				  )
+				: null,
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:53]: What is the benefit of this change?
We are requesting the full entity records just to get the slug instead of the ID. Is the slug essential for the `activeItem`? 🤔

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js [hunk 5]
```diff
@@ -1,80 +1,101 @@
...
 				}
+				onActivateMenu={ setActiveMenu }
 			>
-				<NavigationBackButton
-					backButtonLabel={ __( 'Dashboard' ) }
-					className="edit-site-navigation-panel__back-to-dashboard"
-					href="index.php"
-					ref={ ref }
-				/>
+				{ activeMenu === 'root' && (
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:70]: I'm cool with this change, but I wonder if it might have a11y implications — either way I mean. 🤔

File: packages/edit-site/src/store/actions.js [hunk 6]
```diff
@@ -43,7 +43,19 @@ export function __experimentalSetPreviewDeviceType( deviceType ) {
  *
  * @return {Object} Action object.
  */
-export function setTemplate( templateId ) {
+export function* setTemplate( templateId ) {
+	if ( Number.isNaN( Number( templateId ) ) ) {
```
Comment (Copons) [inline packages/edit-site/src/store/actions.js:47]: Out of curiosity, why do we need such involved check? 😮 
What kind of data could we expect here?
The function docs say that `templateId` is supposed to be a number, so... couldn't we just go with a simple falsy check?

File: packages/edit-site/src/store/controls.js [hunk 7]
```diff
@@ -15,19 +15,24 @@ import { findTemplate as findTemplateUtil } from '../utils';
  *
  * @return {number} The found template ID.
  */
-export function findTemplate( path ) {
+export function findTemplate( path, templateType, templates ) {
```
Comment (Copons) [inline packages/edit-site/src/store/controls.js:18]: Please update the function docs with the new params!

File: packages/edit-site/src/store/actions.js [hunk 8]
```diff
@@ -43,7 +43,19 @@ export function __experimentalSetPreviewDeviceType( deviceType ) {
...
+		templateId = yield findTemplate( '', templateId, [ templateId ] );
+
+		yield dispatch(
+			'core/data',
+			'invalidateResolution',
+			'core',
+			'getEntityRecords',
+			[ 'postType', 'wp_template' ]
+		);
+	}
```
Comment (Copons) [inline packages/edit-site/src/store/actions.js:57]: Mind adding some inline comments to explain what happens here? 🙂

File: packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+} from '@wordpress/components';
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import TemplatePreview from './template-preview';
+import { NavigationPanelPreviewFill } from '.';
+
+const TITLES = {
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:16]: What do you think about moving this constant into its own file?

Comment (Copons) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:16]: Actually, maybe even all constants in this PR could live in a `constants.js` file.
`GENERAL_TEMPLATES`, `GENERAL_TEMPLATE_SLUGS`, and `TITLES` here.
It should make it easier to look up and to update. 🤔

**Comment by david-szabo97** [comment]
> I've noticed a wonky behaviour when changing from template to template part and vice versa.

We have it documented here: https://github.com/WordPress/gutenberg/issues/25743

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js (hunk 5 above)
Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:70]: Two back button is strange, that was the main reason I removed this back button.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js (hunk 4 above)
Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:53]: Thanks for noticing. Reverted it to the ID version. I had to use slug because of a former solution but since then that's resolved.

Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:58]: Reverted it back to the slug version 😬 ID changes after we created an auto-draft, which results in a brief moment when we don't have an active item in the navigation tree. Left a comment in the code to explain it.

File: lib/template-loader.php (hunk 3 above)
Comment (david-szabo97) [reply lib/template-loader.php:101]: We couldn't retrieve templates by specific template types. With these two arguments we can make retrieve a specific template type by specifying the template type and the hierarchy. Hierarchy ($templates) will be the same as $templateType in this case to make sure it returns this specific template, rather than a descendant.

File: packages/edit-site/src/store/actions.js (hunk 6 above)
Comment (david-szabo97) [reply packages/edit-site/src/store/actions.js]: Replaced check with a simple `typeof`. Updates the doc which explains what the function now does.

File: packages/edit-site/src/store/controls.js (hunk 7 above)
Comment (david-szabo97) [reply packages/edit-site/src/store/controls.js:20]: Done! 🧇

File: packages/edit-site/src/store/actions.js (hunk 8 above)
Comment (david-szabo97) [reply packages/edit-site/src/store/actions.js:70]: Actually, this `invalidateResolution` is temporary here. I have no idea where to put this to make more sense.

This function (`setTemplate`) creates an auto-draft if we pass a slug instead of an ID. `invalidateResolution` requires all the args which makes this invalidation very specific. Is there any way to invalidate all the `wp_template` queries? Docs didn't help :(

File: packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js (hunk 9 above)
Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js]: Great idea! Done!

File: lib/edit-site-page-templates.php (hunk 2 above)
Comment (david-szabo97) [reply lib/edit-site-page-templates.php:8]: > what happens with elements with a file, but without an entity (aka without an autodraft)?

Nothing, they are not created automagically, they won't be included in the `getEntityRecords` request. Template parts are created when requested via `getEntityRecords`, but not templates. The auto-draft is only created when we make a specifically request that template. That's why we need to expose an endpoint which returns all the file based templates.

Actually, I think it would be better to implement the same auto-drafting logic for templates. So we don't need two endpoints and it would save us a lot of hassle. I'll look into this.

Comment (david-szabo97) [reply lib/edit-site-page-templates.php]: Replaced existing solution with magic. Templates are now automatically created as auto-drafts when requested. This simplified the code a lot!

**Comment by david-szabo97** [comment]
@vindl @Copons
Made a big change which reverts a lot of code. Ditched my solution regarding the page templates and instead the templates are now automatically created as auto-drafts, just like template parts. This made the code very simple compared to the previous one! I don't see any anomalies or cons to this solution yet.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js (hunk 5 above)
Comment (Copons) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:63]: Right, two back buttons is strange, but at the same time one of them (back to dashboard) is "outside" the navigation menu context.
So, it might not be clear for visually impaired users why that button disappeared while navigating through the menu.

This is all anecdotal, though.
As I said, the change looks good to me, and it's super simple to revert. 🙂

File: packages/edit-site/src/components/left-sidebar/navigation-panel/index.js [hunk 10]
```diff
@@ -56,25 +58,37 @@ const NavigationPanel = () => {
...
+					<TemplatesMenu onActiveIdChange={ setTemplate } />
+
+					<TemplatePartsMenu
 						onActiveTemplatePartIdChange={ setTemplatePart }
-						onAddTemplate={ addTemplate }
-						onRemoveTemplate={ removeTemplate }
 					/>
 				</NavigationMenu>
 			</Navigation>
+
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:90]: Now that we have templates and template parts in separate menus, how about simplifying the API and have `onActivate` (or `onActivateItem`) in both `TemplatesMenu` and `TemplatePartsMenu`, instead of `onActiveIdChange` and `onActiveTemplatePartIdChange`?

File: packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
+export const ITEM_CONTENTS = {
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js:1]: I think we could clarify these constants a little bit.

Something like:
- `ITEM_CONTENTS` -> `TEMPLATES_DEFAULT_DETAILS`
- `GENERAL_TEMPLATES` -> `TEMPLATES_GENERAL` (nitpick)
- `POSTS_GENERAL_TEMPLATES` -> `TEMPLATES_POSTS` (nitpick, but we could lose the "general" here 🤔 )

File: packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	if ( ! templates ) {
+		return null;
+	}
+
+	if ( ! Array.isArray( templates ) ) {
+		templates = [ templates ];
+	}
+
+	const itemPrefix =
+		entityType === 'wp_template_part' ? 'template-part' : 'template';
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:36]: Just realized that we don't really need this "conversion". 🤔 
We could just simply use the template type + the ID.

E.g.
```jsx
item={ `${ entityType }-${ template.id }` }
```

(Here, and in the NavigationPanel as well)

File: packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+						title={ title || template.slug }
+					>
+						<Button
+							onClick={ () => onActivate( template.id ) }
+							onMouseEnter={ () =>
+								onMouseEnterTemplate( template )
+							}
+							onMouseLeave={ onMouseLeaveTemplate }
+						>
+							{ title || template.slug }
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:59]: Wonder if it makes more sense to invert this? 🤔 
`{ template.slug || title }`

File: packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+							{ description && (
+								<div className="edit-site-navigation-panel__template-item-description">
+									{ description }
+								</div>
```
Comment (Copons) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:63]: As [mentioned in another issue](https://github.com/WordPress/gutenberg/issues/25755#issuecomment-702063762), templates _should/could_ allow for user-defined descriptions in the `excerpt` field.

I guess we could do the same we do for the title/slug, and add `template.excerpt` as a fallback — or, as I mentioned in another comment, the "main" field, and treat `description` as fallback.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js (hunk 13 above)
Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:59]: Slug isn't an optional field, so our "default titles" would never be shown. I don't think 'index' (slug) would be more descriptive than "Default (index)".

Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js:59]: Maybe you meant `template.title.rendered`? That would make sense :D In this case we need to make sure it is different from slug.

Comment (Copons) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js]: I was more thinking around the presence of an autodraft (and therefore a slug) or falling back to the original, default template.
Thinking about it, I'm pretty sure I got confused by the whole thing though 😄

Comment (david-szabo97) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/template-navigation-items.js]: We always have an auto-draft now. They are created automatically. Though we would have `slug` even if we didn't have auto-drafts.

**Review by Copons — APPROVED** [review]
This works nice, thanks @david-szabo97!

I'll open two follow up issues shortly:
- Audit and improve the performances of the new auto-draft creation.
- Discuss with design how to handle template overwrite/revert.

File: lib/templates.php [hunk 15]
```diff
@@ -177,6 +202,13 @@ function filter_rest_wp_template_collection_params( $query_params ) {
  * @return array Filtered $args.
  */
 function filter_rest_wp_template_query( $args, $request ) {
+	// Create auto-drafts for each theme template files.
+	$block_template_files = gutenberg_get_template_paths();
```
Comment (vindl) [inline lib/templates.php:206]: As already mentioned, accessing these files on each request is going to have a negative impact on performance. We can probably store this info once on theme activation and we'll also need it after theme updates.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
+export const TEMPLATES_DEFAULT_DETAILS = {
+	// General
+	'front-page': {
+		title: 'Front page',
+		description: '',
```
Comment (vindl) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js:5]: We could've added "Display the contents of your Front Page" here, or something along the similar lines.

File: packages/edit-site/src/components/left-sidebar/navigation-panel/menus/templates-posts.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+export default function TemplatePostsMenu( { templates, onActivateItem } ) {
+	const generalTemplates = templates?.find( ( { slug } ) =>
+		TEMPLATES_POSTS.includes( slug )
+	);
+	const specificTemplates = templates?.filter( ( { slug } ) =>
+		slug.startsWith( 'post-' )
+	);
+
+	return (
+		<NavigationMenu
```
Comment (vindl) [inline packages/edit-site/src/components/left-sidebar/navigation-panel/menus/templates-posts.js:24]: We could've delayed adding this until we had actual content for it.

**Review by vindl — COMMENTED** [review]
Nice work @david-szabo97! ✨

File: packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js (hunk 16 above)
Comment (mcsf) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js:5]: @david-szabo97 @vindl — Let's not forget i18n. Constants files like this should be i18n-ready from the start, otherwise getting everything ready for a testable Site Editor around WP 5.6 will be much more difficult.

Comment (Copons) [reply packages/edit-site/src/components/left-sidebar/navigation-panel/constants.js:5]: @mcsf My fault, I should have caught this in review!
I'll spin up a PR ASAP.
